exercises/alphametics/example.js

```diff
@@ -0,0 +1,107 @@
+function solve(puzzle) {
+  var parts = puzzle
+    .split(/[+|==]/g)
+    .map(function(o) { return o.trim(); })
+    .filter(function(o) { return o !== ""; });
+
+  if(parts.length < 3) {
+    return null;
+  }
+
+  var uniqueLetters = getUniqueLetters(parts.join(''));
+  var firstLetters = getFirstLetters(parts);
+
+  var numberCombinations = getNumberCombinations([0,1,2,3,4,5,6,7,8,9], uniqueLetters.length);
+  var permutations = getPermutations(Array(uniqueLetters.length).fill().map(function(_, i) {return i; }));
+
+  while(numberCombinations.length) {
+    var numberCombination = numberCombinations.pop();
+    for(var k = 0; k < permutations.length; k++) {
+      var newNumbers = assignNumbers(numberCombination, uniqueLetters, permutations[k]);
+      if(testNumbers(newNumbers, parts, firstLetters)) {
+        return newNumbers;
+      }
+    }
+  }
+  return null;
+}
+
+function getFirstLetters(words) {
+  return words
+    .map(function(word) { return word[0]; })
+    .filter(function (val, i, arr) { return arr.indexOf(val) === i; });
+}
+
+function assignNumbers(numbers, letters, orders) {
+  var output = {};
+  for(var i = 0; i < letters.length; i++) {
+    output[letters[i]] = numbers[orders[i]];
+  }
+  return output;
+}
+
+function getUniqueLetters(string) {

```

Comment:
I agree. I plan on converting this to ES6 soon, there are lots of rooms for optimizations.